app/models/user.rb

```diff
@@ -185,13 +185,21 @@ def unread_inbox_comments_count
   scope :valid, -> { where(banned: false, suspended: false) }
   scope :out_of_invites, -> { where(out_of_invites: true) }
 
-  ## used in app/views/users/new.html.erb
-  validates_length_of :login,
-                      within: ArchiveConfig.LOGIN_LENGTH_MIN..ArchiveConfig.LOGIN_LENGTH_MAX,
-                      too_short: ts("^User name is too short (minimum is %{min_login} characters)",
-                                    min_login: ArchiveConfig.LOGIN_LENGTH_MIN),
-                      too_long: ts("^User name is too long (maximum is %{max_login} characters)",
-                                   max_login: ArchiveConfig.LOGIN_LENGTH_MAX)
+  validates :login,
+            length: {
+              within: ArchiveConfig.LOGIN_LENGTH_MIN..ArchiveConfig.LOGIN_LENGTH_MAX,
+              too_short: I18n.t("user.login.too_short", min_login: ArchiveConfig.LOGIN_LENGTH_MIN),
+              too_long: I18n.t("user.login.too_long", max_login: ArchiveConfig.LOGIN_LENGTH_MAX)
+            },
+            format: {
+              with: /\A[A-Za-z0-9]\w*[A-Za-z0-9]\Z/,
+              message: I18n.t("user.login.bad_format",
+                              min_login: ArchiveConfig.LOGIN_LENGTH_MIN,
+                              max_login: ArchiveConfig.LOGIN_LENGTH_MAX)
+            },
```

Comment:
Oh, I forgot to flag this before: the calls to `I18n.t` here aren't inside a block or a method, so they'll be evaluated when the class is loaded, i.e. when the server first boots up. This means that the error messages will always be in the default locale, even if someone is browsing with a different locale set. If you want to make this translatable, you have to pass symbols for each of the messages, or just rely on the default symbols.

(This was also true when using `ts` here. But `ts` means the string isn't translatable yet; whereas `t` ought to mean that translation works.)

The default symbols used by the length validator are `too_short` and `too_long`, and they pass a `count` param so that you can customize the message with e.g. `is too long (maximum is %{count} characters)`. But it's worth noting that I think the default length messages are actually [already set to the desired text](https://github.com/svenfuchs/rails-i18n/blob/556eb65a4b701eec430cfc5954f103a102ca1ae5/rails/locale/en.yml#L134-L139). So it would probably just be best to delete the `too_short` and `too_long` params altogether.

The format error message is a little trickier, because the min and max length aren't normally a part of the format validator. But it looks like you should be able to pass custom parameters to the validator and have them [added as parameters to the error](https://github.com/rails/rails/blob/3779b5069dd1742d878cc0e322fc642e730efa42/activemodel/lib/active_model/validations/format.rb#L31).

Reply:
Ah right, thanks for catching this (and for linking all those classes)! I checked again, and the defaults do indeed seem to be right for everything except `format`, which I have now updated to be the custom message.